databricks/koalas/series.py

```diff
@@ -1913,6 +1914,133 @@ def _fillna(self, value=None, method=None, axis=None, limit=None, part_cols=()):
+        --------
+        >>> kser = ks.Series(['b', None, 'a', 'c', 'b'])
+        >>> codes, uniques = kser.factorize()
+        >>> codes
+        0    1
+        1   -1
+        2    0
+        3    2
+        4    1
+        dtype: int32
+        >>> uniques
+        Index(['a', 'b', 'c'], dtype='object')
+
+        >>> codes, uniques = kser.factorize(na_sentinel=None)
+        >>> codes
+        0    1
+        1    3
+        2    0
+        3    2
+        4    1
+        dtype: int32
+        >>> uniques
+        Index(['a', 'b', 'c', None], dtype='object')
+
+        >>> codes, uniques = kser.factorize(na_sentinel=-2)
+        >>> codes
+        0    1
+        1   -2
+        2    0
+        3    2
+        4    1
+        dtype: int32
+        >>> uniques
+        Index(['a', 'b', 'c'], dtype='object')
+        """
+        assert (na_sentinel is None) or isinstance(na_sentinel, int)
+        assert sort is True
+        uniq_sdf = self._internal.spark_frame.select(self.spark.column).distinct()
+
+        # Check number of uniques and constructs sorted `uniques_list`
+        max_compute_count = get_option("compute.max_rows")
+        if max_compute_count is not None:
+            uniq_pdf = uniq_sdf.limit(max_compute_count + 1).toPandas()
+            if len(uniq_pdf) > max_compute_count:
+                raise ValueError(
+                    "Current Series has more then {0} unique values. "
+                    "Please set 'compute.max_rows' by using 'databricks.koalas.config.set_option' "
+                    "to more than {0} rows. Note that, before changing the "
+                    "'compute.max_rows', this operation is considerably expensive.".format(
+                        max_compute_count
+                    )
+                )
+        else:
+            raise ValueError(
+                "Please set 'compute.max_rows' by using 'databricks.koalas.config.set_option' "
+                "to restrict the total number of unique values of the current Series."
+                "Note that, before changing the 'compute.max_rows', "
+                "this operation is considerably expensive."
+            )
+
+        uniques_list = first_series(uniq_pdf).tolist()
+        uniques_list = sorted(uniques_list, key=lambda x: (pd.isna(x), x))
+
+        # Constructs `unique_to_code` mapping non-na unique to code
+        unique_to_code = {}
+        if na_sentinel is not None:
+            na_sentinel_code = na_sentinel
+        code = 0
+        for unique in uniques_list:
+            if pd.isna(unique):
+                if na_sentinel is None:
+                    na_sentinel_code = code
+            else:
+                unique_to_code[unique] = code
+            code += 1
+
+        kvs = list(
+            chain(*([(F.lit(unique), F.lit(code)) for unique, code in unique_to_code.items()]))
+        )
+
+        map_scol = F.create_map(kvs)
+        null_scol = F.when(self.spark.column.isNull(), F.lit(na_sentinel_code))
```

Comment:
We should handle `NaN` when the data type is `FloatType` or `DoubleType`?

Reply:
May I ask what value in Koalas dataframe would be represented by `NaN` in the Spark dataframe?

Reply:
Spark has `F.isnan` to check whether the value is `NaN` or not. e.g., https://github.com/databricks/koalas/blob/d33d1d3ab6cba12dbcbd79f00eae8b93516a2399/databricks/koalas/series.py#L1876

Reply:
In which case, would `NaN` be in the spark dataframe?
```
>>> kser
0    1.0
1    NaN
2    3.0
3    2.0
4    1.0
dtype: float64

>>> kser._internal.spark_frame.show()
+-----------------+----+-----------------+
|__index_level_0__|   0|__natural_order__|
+-----------------+----+-----------------+
|                0| 1.0|      17179869184|
|                1|null|      34359738368|
|                2| 3.0|      60129542144|
|                3| 2.0|      77309411328|
|                4| 1.0|      94489280512|
+-----------------+----+-----------------+
```

Reply:
e.g.,:

```py
>>> kser = ks.Series(range(5))
>>> kser.loc[3] = np.nan
>>> kser._internal.resolved_copy.spark_frame.show()
+-----------------+---+-----------------+
|__index_level_0__|  0|__natural_order__|
+-----------------+---+-----------------+
|                0|0.0|      17179869184|
|                1|1.0|      34359738368|
|                2|2.0|      60129542144|
|                3|NaN|      77309411328|
|                4|4.0|      94489280512|
+-----------------+---+-----------------+
```

Reply:
Also please don't forget that Koalas objects can be converted from existing Spark DataFrames which can include `NaN`s.

Reply:
For the `kser.loc[3] = np.nan` example, is `NaN` in the spark_frame as a result of `resolved_copy`?

Reply:
`kser.loc[3]` doesn't update the internal Spark DataFrame but only update the `data_spark_columns`. We need to use `resolved_copy` to see the actual change.

Reply:
Got it, thanks!